### ISSUE TITLE
Make ZStream#drop and ZStream#take more intuitive when handling errors

### ISSUE THREAD

**Issue by vasilmkd** [zio/zio#2525]
With element level errors, the users will be able to retry a failed pull on a stream. How should `drop` and `take` handle these situations? Should failures be counted as dropped and taken elements, and it would be up to the user to determine how they want to replace the failed pulls with values, or should only the successful pulls be counted as dropped and taken elements?

Currently, drop and take behave differently in this regard and we need to make them consistent. For an implementation and test results that follow the second approach, please refer to #2524 .

@iravid @jdegoes @regiskuckaertz @adamgfraser @mschuwalow and anyone interested is welcome to join the discussion.

**Comment by jdegoes** [comment]
A useful exercise when faced with such design questions: Which semantic preserves user choice, i.e. allows deriving the other semantic given other appropriate operators? Further, if both semantics preserve user choice, which one is least surprising in edge cases? Finally, which semantic can help the operations participate in new and useful algebraic laws?

Let's imagine that `take` will count errors toward its quota. Then how can we get `take` to **not** count errors toward its quota? It seems to me we need to materialize to a stream of eithers, and then collect only the rights, and then `take`.

Now let's imagine that `take` will **not** count errors toward its quota. Then how can we get `take` **to** count errors toward its quota? We need to materialize to a stream of eithers, and then `take`.

If I haven't made any mistakes, you can see we can achieve both semantics through either choice. However, the second choice — `take` does not count errors toward its quota — has a much more straightforward solution, calling to mind `race`, which can achieve "give me the first completion" by `left.either race right.either`.

Now onto the second question: which one has is least surprising in edge cases?

Well, I can only think of one "edge case": let's say we had a stream of failures concatenated with a stream of successes, e.g. `failures ++ successes`. The left stream has size 0, but generates 100 failures. The right stream has size 100. It would be kind of surprising (to me) that if you `drop` 100 elements, you end up with a stream that succeeds with 100 elements—namely, the only elements that were ever in the original stream. 

You can also imagine us having a "recovery" operator that can map an original failure in the source stream into `n` failures in the output stream (e.g. it retries a certain number of times but preserves all those failures). This would similarly lead to surprising behavior around element counts.

Surprises like these are often "law violations", which brings us to the third question: which semantic can more easily participate in laws?

Now size on effectful streams is not defined in general, because it's nondeterministic. But if you imagine it were defined, then you would certainly like that:

```scala
stream.take(stream.size) == stream
stream.drop(stream.size) == Stream.empty
```

And so forth. And while it is not possible to define size for effectful streams, one can define it for "pure streams" and require these relations hold. You could even go one step beyond that and define a corresponding version for nondeterministic streams (although it's quite tricky).

What do all these considerations suggest?

**Comment by vasilmkd** [comment]
Amazing explanation @jdegoes. Thank you for your time.

It seems failures will count towards the quota, and `drop` is the one that needs rewriting. Thinking about it, when I tackled this yesterday, `drop`'s not counting towards the quota was initially more surprising to me, but for some reason I went with the flow which led me to a mistake.

Thanks again, I will amend the PR.

**Comment by regiskuckaertz** [comment]
> It seems failures will count towards the quota

🤔 I think John was pointing at the opposite, i.e. `(Stream.fail("Boom") ++ Stream(1,2,3)).drop(3) == Stream.empty`?

**Comment by regiskuckaertz** [comment]
woops, sorry ☝️

**Comment by vasilmkd** [comment]
I'm even more confused now.

**Comment by vasilmkd** [comment]
Can someone please take a look at the test code in #2524 and decide if it makes sense?

**Comment by regiskuckaertz** [comment]
> I'm even more confused now

With that version, you can get the other by just doing (Stream.fail("Boom") ++ Stream(1,2,3)).either.drop(3).absolve == Stream(3)` (`absolve` is currently missing)

> Now I feel stupid

😨 not. at. all. If I read that correctly, you only want to increase the counter if `as` succeeds, but that cannot be expressed with `ref.modify`. I think this is correct, the only missing ingredient would be forgetting any error, right?

**Comment by vasilmkd** [comment]
I already did it in my original implementation of #2524 but I couldn't explain my reasoning to @iravid, that's why I feel bad now. Please check it out @regiskuckaertz. Thank you.

**Comment by jdegoes** [comment]
@vasilmkd 

Essentially, what I am proposing is to have `take` only take the first `n` successes, and to have `drop` only drop the first `n` successes. Then you can achieve the other semantic with `either` in a straightforward fashion as @regiskuckaertz proposed, and, moreover, with this semantic, we can ensure that spurious effectful failures are not counted toward "stream size", which can help avoid unintended surprises.

Does this make sense?

**Comment by vasilmkd** [comment]
That's what I implemented in #2524, but I get the feeling that @regiskuckaertz and @iravid actually consider this to be _less_ intuitive for some reason.

**Comment by iravid** [comment]
I’m undecided and can see arguments going both ways. After thinking about this today I think I’m happy to go ahead with what you implemented @vasilmkd. Worst case we can always change later in a binary compatible fashion.

**Comment by vasilmkd** [comment]
For some reason this feels like one of the most stressful issues I've tackled in ZIO. 😅

**Comment by iravid** [comment]
Fixed in #2524 - thanks @vasilmkd!
